src/linear_algebra/char_poly/coeff.lean

```diff
@@ -0,0 +1,191 @@
+
+## Main definitions
+
+- `degree_char_poly_eq_dim`
+- `det_from_char_poly`
+- `trace_from_char_poly`
+
+## Implementation details
+
+-/
+
+noncomputable theory
+
+universes u v w z
+
+open polynomial matrix
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R]
+variables {n : Type v} [fintype n] [decidable_eq n]
+variables {α : Type v} [decidable_eq α]
+
+open finset
+open polynomial
+
+section fixed_points
+-- TODO: move this section to be with other combinatorial facts about the symmetric group
+lemma gt_one_nonfixed_point_of_nonrefl {σ : equiv.perm n} (h : σ ≠ equiv.refl n) :
+1 < (filter (λ (x : n), ¬ σ x = x) univ).card :=
+begin
+  rw one_lt_card_iff,
+  contrapose! h, ext, dsimp,
+  have := h (σ x) x, simp only [true_and, mem_filter, equiv.apply_eq_iff_eq, mem_univ, ne.def] at this,
+  tauto,
+end
+
+lemma lt_card_sub_one_fixed_point_of_nonrefl {σ : equiv.perm n} (h : σ ≠ equiv.refl n) :
+(filter (λ (x : n), σ x = x) univ).card < fintype.card n - 1:=
+begin
+  have hun := @filter_union_filter_neg_eq _ (λ (x : n), σ x = x) _ _ _ univ,
+  have hin : (filter (λ (x : n), σ x = x) univ) ∩ (filter (λ (x : n), ¬ σ x = x) univ) = ∅
+    := filter_inter_filter_neg_eq univ,
+  rw ← disjoint_iff_inter_eq_empty at hin,
+  rw fintype.card, conv_rhs { rw ← hun },
+  rw card_disjoint_union hin,
+  have := gt_one_nonfixed_point_of_nonrefl h, omega,
+end
+
+end fixed_points
+
+variable {M : matrix n n R}
+
+lemma nat_degree_char_matrix_val [nontrivial R] (i j : n) :
+  (char_matrix M i j).nat_degree = ite (i = j) 1 0 :=
+by { by_cases i = j; simp [h, ← degree_eq_iff_nat_degree_eq_of_pos (nat.succ_pos 0)], }
+
+lemma nat_degree_char_matrix_val_le (i j : n) :
+  (char_matrix M i j).nat_degree ≤ ite (i = j) 1 0 :=
+by split_ifs; simp [h, nat_degree_X_sub_C_le]
+
+variable (M)
+lemma low_degree_char_poly_sub_diagonal :
+(char_poly M - ∏ (i : n), (X - C (M i i))).degree < ↑(fintype.card n - 1) :=
+begin
+  rw [char_poly, det, ← insert_erase (mem_univ (equiv.refl n)),
+    sum_insert (not_mem_erase (equiv.refl n) univ), add_comm],
+  simp only [char_matrix_apply_eq, one_mul, equiv.perm.sign_refl, id.def, int.cast_one,
+    units.coe_one, add_sub_cancel, equiv.coe_refl],
+  rw ← mem_degree_lt, apply submodule.sum_mem (degree_lt R (fintype.card n - 1)),
+  intros c hc, rw [← C_eq_int_cast, C_mul'],
+  apply submodule.smul_mem (degree_lt R (fintype.card n - 1)) ↑↑(equiv.perm.sign c),
+  rw mem_degree_lt, apply lt_of_le_of_lt degree_le_nat_degree _, rw with_bot.coe_lt_coe,
+  apply lt_of_le_of_lt _ (lt_card_sub_one_fixed_point_of_nonrefl (ne_of_mem_erase hc)),
+  apply le_trans (polynomial.nat_degree_prod_le univ (λ i : n, (char_matrix M (c i) i))) _,
+  rw card_eq_sum_ones, rw sum_filter, apply sum_le_sum,
+  intros, apply nat_degree_char_matrix_val_le,
+end
+
+lemma high_coeff_char_poly_eq_coeff_prod_diag {k : ℕ} (h : fintype.card n - 1 ≤ k) :
+  (char_poly M).coeff k = (univ.prod (λ (i : n), X - C (M i i))).coeff k :=
```

Comment:
Use big operator notation?

Reply:
I don't like the `high_` in the name, I would rather have a `_of_le` suffix.
Reason: if you start searching for `coeff_char_poly` then you want autocompletion to do the rest for you. People will not know to start the lemma name with `high_`.